### PR TITLE
Implement Huber loss

### LINKED ISSUE #1441: Problematic derivative of Tensor::abs and Huber loss
Implementing the Huber loss requires comparing the absolute value against some small `kappa`, then behaving linearly with the absolute value outside this bound, but quadratic inside. An example implementation would be:

```rust
fn huber_loss<const D: usize, B: Backend>(kappa: f32, delta: Tensor<B, D>) -> Tensor<B, D> {
    let u = delta.clone().abs();
    let is_small = u.clone().lower_elem(kappa);
    let inside = delta.powf_scalar(2.).mul_scalar(0.5);
    let outside = u.mul_scalar(kappa).sub_scalar( (kappa * kappa) / 2. );
    outside.mask_where(is_small, inside)
}
```

Using this implementation leads to `NaN` values showing up after a few steps of training. This is probably connected to the gradient computation in `Tensor::abs` but as the problematic small values should be masked out and instead use the (perfectly fine) gradient computation from the quadratic term, I'm not sure how that is happening.

### PR DESCRIPTION
**Pull request by WorldSEnder**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Closes #1441 as I think the remaining feature request for a sign function is already tracked in #522.

### Changes

Implements the [Huber Loss function](https://en.wikipedia.org/wiki/Huber_loss).

Instead of strictly following the definition of using a sign or abs function, the implementation uses clamping, which computes the same value outside the `delta` bounds but is better behaved on the autodiff backend and does not need any extra primitive ops. See also #1441 for my first attempt of implementing this.

### Testing

Test data should cover all relevant branches of the operation, and critical points on the autodiff backend, i.e. zero residuals and the point where the loss switches between the branches. Test assertions have been generated from executing the equivalent in scipy.

Note: the `test_downsample_interpolation` test in `nearest_interpolate.rs` is failing locally for me. Not caused by the patch, I've ignored it when running run-checks.

### CODE REVIEW

**Comment by antimora** [comment]
Submitted a PR for `sign` tensor operator: https://github.com/tracel-ai/burn/pull/1446

**Comment by WorldSEnder** [comment]
Note: I think the method of `clamp`ing even works out better for Huber than using `sign`, since it avoids a `mul_scalar(delta)`. Though I'm still interested in `sign` for alternative losses such as SmoothL1.

**Comment by antimora** [comment]
CI failed due to:

```
 test tests::jit_fusion::var::tests::test_var_mean_bias ... ok
  
  failures:
  
  ---- tests::jit::kernel::normal::tests::subsequent_calls_give_different_tensors stdout ----
  thread 'tests::jit::kernel::normal::tests::subsequent_calls_give_different_tensors' panicked at crates/burn-wgpu/src/lib.rs:73:5:
  assertion failed: tensor_1.to_data().value[i] != tensor_2.to_data().value[i]
  
  
  failures:
      tests::jit::kernel::normal::tests::subsequent_calls_give_different_tensors
  
  test result: FAILED. 1341 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 55.86s
```

Rerunning to see if it's a fluke. Tagging @nathanielsimard and @louisfd since they're currently working in this area.

**Comment by antimora** [comment]
Sign tensor op PR is merged.

**Review by louisfd — APPROVED** [review]
LGTM, but can you clarify the math in comments? I think it works out but it's a bit confusing, in particular I think `r`, `err` and `res` are three names for the same thing?
